framework/include/libtorch/utils/LibtorchArtificialNeuralNetTrainer.h

```diff
@@ -0,0 +1,318 @@
+ * of this class is that it can be used with random/sequential samplers in serial or
+ * parallel. The default sampling approach is sequential.
+ */
+template <typename SamplerType = torch::data::samplers::DistributedSequentialSampler>
+class LibtorchArtificialNeuralNetTrainer : public libMesh::ParallelObject
+{
+public:
+  /**
+   * Construct using the neural network and a parallel communicator
+   * @param nn The neural network which needs to be trained
+   * @param comm Reference to the parallel communicator
+   */
+  LibtorchArtificialNeuralNetTrainer(std::shared_ptr<LibtorchArtificialNeuralNet> nn,
+                                     const Parallel::Communicator & comm);
+
+  /**
+   * Train the neural network using a given (serialized) data and options for the training process
+   * @param dataset The dataset containing the known input-output pairs for the training
+   * @param options Options for the optimizer/training process
+   */
+  virtual void train(LibtorchDataset & dataset, const LibtorchTrainingOptions & options);
+
+  /**
+   * Computes the number of samples used for each batch.
+   * @param num_samples The total number of samples used for the training
+   * @param num_batches The total number of batches we want to select. This is just a target number,
+   * based on load-balancing considerations, the code can deviate from this by 1
+   * @return The number of samples in the batches
+   */
+  static unsigned int computeBatchSize(const unsigned int num_samples,
+                                       const unsigned int num_batches);
+  /**
+   * Computes the number of local samples. This practically splits the already computed batches
+   * between MPI ranks. At the moment, this is a very simple logic but will change in the future
+   * when custom dataloaders are added.
+   * @param batch_size The batch size which needs to be split
+   * @param num_ranks The number of ranks associated with this batch
+   * @return The number of samples in a batch which will be used locally to train the neural nets
+   */
+  static unsigned int computeLocalBatchSize(const unsigned int batch_size,
+                                            const unsigned int num_ranks);
+
+  /**
+   * Setup the optimizer based on the provided options.
+   * @param nn The neural network whose parameters need to be optimized
+   * @param options The options for the training process
+   */
+  static std::unique_ptr<torch::optim::Optimizer>
+  createOptimizer(const std::shared_ptr<LibtorchArtificialNeuralNet> nn,
+                  const LibtorchTrainingOptions & options);
+
+protected:
+  /// Pointer to the neural network which is trained
+  const std::shared_ptr<LibtorchArtificialNeuralNet> _nn;
+};
+
+template <typename SamplerType>
+LibtorchArtificialNeuralNetTrainer<SamplerType>::LibtorchArtificialNeuralNetTrainer(
+    std::shared_ptr<LibtorchArtificialNeuralNet> nn, const Parallel::Communicator & comm)
```

Comment:
does this need to be a shared_ptr?

Reply:
Note, this guy decided to be a reference starting from the next commit.